src/librustc/middle/effect.rs

```diff
@@ -178,11 +177,28 @@ impl<'a, 'tcx, 'v> Visitor<'v> for EffectCheckVisitor<'a, 'tcx> {
                     self.require_unsafe(expr.span, "use of mutable static");
                 }
             }
+            hir::ExprField(ref base_expr, field) => {
+                if let ty::TyUnion(..) = self.tcx.expr_ty_adjusted(base_expr).sty {
```

Comment:
Nit: I feel like it'd be nice to `match` the sty (exhaustively) and have a `bug!` call for the unexpected cases. This makes me mildly nervous that some bug or addition might slip through.